### PR TITLE
Exchange edges

### PR DESCRIPTION
**Pull request by dgtony**
Follow-up to #477.

Here is a rework of internal node communication patterns. It should significantly reduce traffic and CPU consumption of the nodes in a steady state and make it scale better. Surely, the PR is huge, but unfortunately it turned out to be impractical to divide it into the smaller ones, as all the features play well just being implemented together.

Here is the outline:
* Internal communication protocol for ThreadsDB nodes extended with a new call `ExchangeEdges`. New method enables efficient checks if thread information is in sync between any two nodes, using heads/address edges (digest), and is a replacement for continuous pulling with `GetRecords`. It also helps to reduce communication overhead with both synced and inactive peers.
* Now all calls to `GetLogs`/`GetRecords` are grouped by peers and performed through the corresponding queues, thus deduplicating and distributing peer calls over the time.
* Address edge computation implemented - deterministic digest of all peer addresses involved into the thread (similar to heads edge).
* Multiple minor fixes and improvements.

New communication logic is fully compatible with an older nodes, as it fallbacks to the `GetRecords`-based pulling if peer is lacking support of `ExchangeEdges`.

### CODE REVIEW

**Comment by sanderpick** [comment]
Really nice @dgtony! I gave it a skim, but would you mind adding some annotations here and there throughout the code to help me follow the big pieces? Easiest way to do that is by doing a self-review w/comments.

File: net/server.go
Comment (dgtony) [inline net/server.go:331]: we don't know about the thread - let's create it (later)

File: logstore/lstoreds/headbook.go
Comment (dgtony) [inline logstore/lstoreds/headbook.go:167]: large instances would better use hybrid logstore with edges fetched from the in-memory storage to avoid performance issues

File: logstore/lstoremem/addr_book.go
Comment (dgtony) [inline logstore/lstoremem/addr_book.go:371]: similar to `HeadsEdges` approach - edges in the in-memory implementation are recomputed on the thread's info change as its pretty cheap operation here

File: net/client.go
Comment (dgtony) [inline net/client.go:149]: synchronous call

File: net/queue/ff.go
Comment (dgtony) [inline net/queue/ff.go:122]: receiving, keeping and executing calls evenly and in time for the multiple peers

File: net/server.go
Comment (dgtony) [inline net/server.go:231]: save traffic and CPU if log is empty

File: net/net.go
Comment (sanderpick) [inline net/net.go:1232]: It would be great to distribute this over some time interval by paginating over the threads in an efficient manner. When there are 10s of 1000s of threads, this is a pretty heavy call.

**Review by sanderpick — APPROVED** [review]
Hey @dgtony, apologies for the delay here. This is a huge contribution... thank you! I left a couple comments and questions, but I'm happy to merge as is. @carsonfarmer can you give this a skim when you have a moment just so we're all on the same page about the new functionality? 

Also, I'd love to hear where the scaling bottlenecks are for your application. I have some ideas to improve things, but your usage feedback would be really helpful in deciding what direction to go. Hopefully we can have the team call that was mentioned over email soon.

File: common/common.go
Comment (dgtony) [reply common/common.go:62]: It's already added in function `persistentStore`

File: net/queue/ff.go
Comment (dgtony) [reply net/queue/ff.go:242]: As the reason of the issue is contention on the thread's head-keys, I believe that optimal solution depends on the intensity of writes to the logs. While contention is relatively low, I'd rather use optimistic approach with retried transactions. In case of unusually intensive writes and high contention pessimistic locking/semaphores would perform better.

Comment (dgtony) [reply net/queue/ff.go:242]: Regarding queue polling, that's a tricky part )
Ok, so whats going on. Here we have an underlying FIFO-queue containing every scheduled call associated with the moment it was added in the linked list. There is a predefined parameter `timeout` - every scheduled call should be spawned during such timeout, though it's not a hard deadline, just a desired property. And moreover, we want it to be as uniform over time as possible (in the presense of non-stationary process of new call arrivals).
On every polling iteration we repeatedly make a decision: pop and spawn next call from the queue or wait until the next polling iteraton. It'd be too expensive to obtain precise deadline distribution from the large queue, but since it's FIFO we know at least they are in the interval `[last_popped_deadline; timeout]`. And, of course, number of calls waiting in the queue is known as well.
Not much information, but we can use some randomized model for polling decisions and it may help to avoid synchronization with other queues. So here is the equation! It doesn't have solid theoretical underpinning, rather it's more an empirical one, reflecting two basic assumptions: probability of taking a call should be: 1) inversely proportional to the worst case deadline (when all calls in the queue share the same deadline) and 2) sublinear of a queue size.
So far its the best model I've been able to come up with in terms of smooth operation and meeting deadlines according to both our simulations and real app metrics. But I'll be happy to replace it with a more solid-ground algorithm, if any.

File: logstore/lstoreds/headbook.go
Comment (dgtony) [reply logstore/lstoreds/headbook.go:167]: As the reason of the issue is contention on the thread's head-keys, I believe that optimal solution depends on the intensity of writes to the logs. While contention is relatively low, I'd rather use optimistic approach with retried transactions. In case of unusually intensive writes and high contention pessimistic locking/semaphores would perform better.

File: net/queue/ff.go
Comment (sanderpick) [reply net/queue/ff.go:242]: Thanks for the explanation! Happy to defer to you here. We can play with it during Testground testing.

Comment (eightysteele) [reply net/queue/ff.go:242]: (👋🏻 @dgtony really great context here—small suggestion would be to include it in a comment for future readers!)

**Comment by carsonfarmer** [comment]
Ok folks, I've given this a thrice-over, but to be honest, I didn't do a deep dive on all the updates here. I'm also happy to merge mostly as is, though I do really like @eightysteele's suggestion of some doc strings that outline the algorithm used in net/queue/ff.go, so I'd like to see that updated before we merge. Otherwise, I'm keen to just play around with this and see how it behaves.

File: net/queue/ff.go
Comment (dgtony) [reply net/queue/ff.go:264]: added comment and renamed vars for clarity

**Pull request merged**
